### DISCUSSION TITLE
[Feature request] Keep notifications always on top

### DISCUSSION THREAD

**Discussion by 4164696f73** [dunst-project/dunst#1254] [General]
TLDR: Dunst's Fullscreen option (I believe it was "pushback") works just like that already - if a fullscreen window is focused, the notification is hidden, if you swap to a non-fullscreen window, the notification pops on top of the fullscreen window. I'd like this popping on top behavior to be possible with all windows at all times (obviously it being optional, not enforced).

**Some more bullshit and explanation and things you probably don't care about**:
I've been trying to figure this out on and off, with Dunst or my Dwm config, for a few months, but couldn't figure it out so decided to post this feature request.

Dunst works all fine in general, with the exception of switching focus and triggering XRaiseWindow, which I use in my Dwm for I assume obvious reasons. My issue is that sometimes, even with fullscreen settings (in Dunst) set to different values, notifications do not show on fullscreened windows, or when I switch window focus, the window is placed on top of Dunst, hiding the notifications. 
I find myself being unable to change it in any way, even in Dwm, and even if it's possible through WMs, would be kind of hard to adjust for other WMs that don't offer extensible configuration like Awesome, Dwm, and would require you to modify the source code and modify it every new WM release.

The main reason this would be useful is because I sometimes do not see any notifications, and the dunstctl history log shows even hundreds of them sometimes. Just yesterday, I thought my Dunst crashed or something else happened that caused me to not see any notifications - 117 notifications in dunstctl history (I do not believe Dunst just refused to show them!).

**Comment by fwsmit** [comment]
I think you are looking for the option `fullscreen=show`. This should be the default value.

**Comment by 4164696f73** [comment]
I do have that enabled. But even so, if I switch to another client, the notification is hidden behind the window I've swapped to. So basically, if I have two tiled windows, not in fullscreen, and I swap to the one that has a notification on top of it, the window goes above the notification, hiding it. The same actually happens in fullscreen windows - sometimes even fullscreen=show does nothing here.

**Comment by fwsmit** [comment]
This has to be investigated more to be actionable. I'll convert it to a discussion

**Comment by bynect** [comment]
To solve this probably is needed a check that brings the window on top if anything is moved above

**Comment by eeeXun** [comment]
Any updates here 👀

**Reply by bynect** [reply]
If this is still relevant I will try some solutions

**Reply by eeeXun** [reply]
I've tested it on DE, i.e. mate desktop without this problem.
This problem only occurs on WM with picom as compositor. Is this problem relevant to compositor?

**Reply by bynect** [reply]
I don't think picom is the culprit, but have you tried if it happens when disabled?
The problem is probably that wm like dwm do not implement the ewmh specs and so the dunst window is not pinned above

If you have some specific examples of what happens / what should happen I will look into them

**Reply by eeeXun** [reply]
> I don't think picom is the culprit, but have you tried if it happens when disabled?

Yes, you are right. Disabling picom didn't fix the problem.

> If you have some specific examples of what happens / what should happen I will look into them

Below is the example in dwm

step1: Send notifications via terminal
step2: Make terminal fullscreen, and the dunst is under the terminal. The expected behavior is that dunst should above the terminal. (This is not limited to terminal, it can be browser, video player, etc.)

https://github.com/user-attachments/assets/2d6b42a2-88b6-4635-ae93-c5cc145c0f7c
